src/gui/programupdater.cpp

```diff
@@ -163,10 +175,55 @@ void ProgramUpdater::rssDownloadFinished(const Net::DownloadResult &result)
         }
     }
 
-    emit updateCheckFinished();
+    if (m_oneReqCompleted)
+        emit updateCheckFinished();
+    else
+        m_oneReqCompleted = true;
+}
+
+void ProgramUpdater::fallbackDownloadFinished(const Net::DownloadResult &result)
+{
+    if (result.status != Net::DownloadStatus::Success)
+    {
+        qDebug() << "Downloading the qBittorrent versions JSON failed:" << result.errorString;
+        if (m_oneReqCompleted)
+            emit updateCheckFinished();
+        else
+            m_oneReqCompleted = true;
+        return;
+    }
+
+    qDebug("Finished downloading qBittorrent versions JSON");
+
+    const auto json = QJsonDocument::fromJson(result.data);
+
+#if defined(Q_OS_MACOS)
+    if (const auto verJSON = json[u"macos"_s][u"version"_s]; verJSON.isString()) {
+#elif defined(Q_OS_WIN)
+    if (const auto verJSON = json[u"win"_s][u"version"_s]; verJSON.isString()) {
+#endif
+        const auto ver = verJSON.toString();
+        if (isVersionMoreRecent(ver))
+            m_fallbackVersion = ver;
+    }
+
+    if (m_oneReqCompleted)
+        emit updateCheckFinished();
+    else
+        m_oneReqCompleted = true;
 }
 
 bool ProgramUpdater::updateProgram() const
 {
-    return QDesktopServices::openUrl(m_updateURL);
+    return shouldUseFallback() ? QDesktopServices::openUrl(u"https://www.qbittorrent.org/download"_s) : QDesktopServices::openUrl(m_updateURL);
 }
+
+bool ProgramUpdater::shouldUseFallback() const
```

Comment:
Seems this function can be called multiple times for the same versions. Wouldn't it be better to parse versions once they are received and store them as `Version` values?